feat!: implement operator authorization policy management (#1384)
defenseunicorns/uds-core: docs/reference/configuration/authorization-policies.md

---
title: Authorization Policies in Ambient Mode
sidebar:
  order: 2
---

## Overview

> In an ambient mode service mesh environment, security is enforced using Istio AuthorizationPolicies. The UDS‑Core operator automatically generates these **ALLOW** policies based on your UDSPackage configuration. The generated policies ensure that only explicitly permitted traffic is allowed, enforcing per‑port access control for your applications. In addition to handling custom ingress rules (allow/expose), the operator also processes monitor configurations to protect monitoring endpoints (such as Prometheus metrics).

Reviewer: ```suggestion
In an ambient mode service mesh environment, ingress network security is partially enforced using Istio AuthorizationPolicies. The UDS‑Core operator automatically generates these **ALLOW** policies based on your UDSPackage configuration. The generated policies ensure that only explicitly permitted traffic is allowed, enforcing per‑port access control for your applications. In addition to handling custom ingress rules (allow/expose), the operator also processes monitor configurations to protect monitoring endpoints (such as Prometheus metrics).
```

Feel free to change wording, just want to note that this is ingress network security essentially.

      targetPort: 3000
```

**Generated AuthorizationPolicy:**
```yaml
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: protect-grafana-ingress-grafana-istio-tenant-gateway
  namespace: grafana
  labels:
    uds/package: grafana
    uds/generation: "1"
spec:
  action: ALLOW
  selector:
    matchLabels:
      app.kubernetes.io/name: grafana
  rules:
    - from:
        - source:
            namespaces: ["monitoring"]
      to:
        - operation:
            ports: ["3000"]
```
*Note: Each monitor rule produces its own policy to restrict access from the monitoring namespace.*

## Summary
- The UDS‑Core operator automatically generates ALLOW Istio AuthorizationPolicies based on your UDSPackage configuration.

- Policies are generated for rules defined in the allow, expose, and monitor blocks, with each rule processed individually.

- Allow rules determine the source based on remote namespace settings, remoteGenerated, and remoteServiceAccount (which specifies a principal).

- Expose rules set the source using principals based on the specified gateway.

- Monitor rules produce individual policies to secure monitoring endpoints.

> - Generated policy names incorporate an "ingress-" naming pattern along with metadata for traceability.

Reviewer: Open to other thoughts but I wonder if most of the content in this doc fits better as a dev doc? I don't think end users generally want/need to know how we translate the `Package` CR spec to other resources (we don't typically document this for servicemonitor, virtualservice for example).

For end user facing documentation I do see a few things that would definitely still be valuable:
- network security is enforced primarily with network policies
- to support istio ambient we also enforce network security (for ingress) with ALLOW authorization policies
- as a result, end user authorization policies should be implemented as DENY policies (could link to how Istio evaluates rules)
- for "best" security we recommend adding `remoteServiceAccount`

Author: Yeah i agree with this, i had made a note of that and forgot to add it in the PR.

---
title: How Authorization Policies Protect Your Services
sidebar:
  order: 3
---

In clusters running Istio Ambient Mesh, UDS‑Core enforces **ingress network security** using Istio **ALLOW** AuthorizationPolicies. These policies are automatically generated for each application package you define with a UDSPackage resource.

> 

Reviewer 2: ```suggestion
[UDS Package](https://uds.defenseunicorns.com/reference/configuration/uds-operator/package/)
```
I think we should try to be consistent with our naming of the UDS Package CR across our documentation. In the codebase we refer to UDS Packages as a `UDSPackage`, but it might not be immediately clear to someone when reading the docs that we are referring to the Package Custom Resource for the operator.

This document explains what this means for you as an application developer and how to take full advantage of the built-in security model.

---

## Key Takeaways

- **Ingress is denied by default.** UDS-Core only allows what you explicitly configure using `allow` and `expose` rules.

- **AuthorizationPolicies are ALLOW-based**, which means you must write **DENY** rules separately if you want to restrict internal traffic further.

- **Use `remoteServiceAccount` wherever possible.** This provides the most secure and identity-based access control.

- **Expose rules use gateways** to control what traffic enters your application. You can choose between:
  - **Tenant Gateway** (default)
  - **Admin Gateway** (used only when absolutely necessary)

- **Monitoring ports are automatically secured** using rules that only allow the `monitoring` namespace to scrape metrics.

---

## Best Practices for Secure Configuration

### 1. Lock Down Ingress With `allow`

```yaml
spec:
  network:
    allow:
      - direction: Ingress
        remoteNamespace: "external-app"
        remoteServiceAccount: "my-client"
        port: 8080
```

> > This ensures that only a specific service account in another namespace can access your service.

Reviewer 2: ```suggestion
> This ensures that only a workload running as this specific service account in another namespace can access your service.
```

### 2. Expose Your Service Safely

```yaml
spec:
  network:
    expose:
      - port: 80
        targetPort: 8080
        gateway: Tenant
```

> This exposes your service at port 80 through the tenant gateway and maps it to your app’s port 8080.

### 3. Enable Safe Monitoring

```yaml
spec:
  monitor:
    - targetPort: 3000
      selector:
        app.kubernetes.io/name: grafana
```

> This creates a rule that allows only Prometheus (from the `monitoring` namespace) to scrape your service.

---

## Authservice Guidance

If you're using Authservice, be aware that it is **only appropriate for simple access scenarios**, such as:

- Protecting web UIs or dashboards
- Cases where access can be fully granted or denied with no granularity

---

## How Istio Evaluates Policies

Istio checks **DENY policies first**, then **ALLOW policies**.

- The operator creates ALLOW policies to admit approved ingress traffic.
- You should create your own DENY policies for more fine-grained control.

More info: [Istio Authorization Policy Evaluation](https://istio.io/latest/docs/concepts/security/#authorization-policy)

---

## Summary

- Ingress is denied by default.

> - You allow ingress by defining `allow` or `expose` rules in your UDSPackage.

Reviewer 2: ```suggestion
- You allow ingress by defining `allow` or `expose` rules in your UDS Package resource definition.
```
